src/components/shared/thread/ActivityGraphCanvas.js

```diff
@@ -0,0 +1,224 @@
+// @flow
+
+import * as React from 'react';
+import { computeActivityGraphFills } from './ActivityGraphFills';
+import { timeCode } from '../../../utils/time-code';
+import { mapCategoryColorNameToStyles } from '../../../utils/colors';
+
+import type {
+  Thread,
+  Milliseconds,
+  SelectedState,
+  IndexIntoSamplesTable,
+  CategoryList,
+} from 'firefox-profiler/types';
+
+import type { CategoryDrawStyles } from './ActivityGraphFills';
+
+type CanvasProps = {|
+  +className: string,
+  +trackName: string,
+  +fullThread: Thread,
+  +interval: Milliseconds,
+  +rangeStart: Milliseconds,
+  +rangeEnd: Milliseconds,
+  +samplesSelectedStates: null | SelectedState[],
+  +treeOrderSampleComparator: (
+    IndexIntoSamplesTable,
+    IndexIntoSamplesTable
+  ) => number,
+  +categories: CategoryList,
+  +passFillsQuerier: any,
+  +onMouseUp: any,
+|};
+
+export class ActivityGraphCanvas extends React.PureComponent<CanvasProps> {
+  _canvas: {| current: null | HTMLCanvasElement |} = React.createRef();
+  _categoryDrawStyles: null | CategoryDrawStyles = null;
+
+  _renderCanvas() {
+    const canvas = this._canvas.current;
+    if (canvas !== null) {
+      timeCode('ThreadActivityGraph render', () => {
+        this.drawCanvas(canvas);
+      });
+    }
+  }
+
+  /**
+   * Get or lazily create the category info. It requires the 2d ctx to exist in order
+   * to create the fill patterns.
+   */
+  _getCategoryDrawStyles(ctx: CanvasRenderingContext2D): CategoryDrawStyles {
+    if (this._categoryDrawStyles === null) {
+      // Lazily initialize this list.
+      this._categoryDrawStyles = this.props.categories.map(
+        ({ color: colorName }, categoryIndex) => {
+          const styles = mapCategoryColorNameToStyles(colorName);
+          return {
+            ...styles,
+            category: categoryIndex,
+            filteredOutByTransformFillStyle: _createDiagonalStripePattern(
+              ctx,
+              styles.unselectedFillStyle
+            ),
+          };
+        }
+      );
+    }
+
+    return this._categoryDrawStyles;
+  }
+
+  componentDidMount() {
+    this._renderCanvas();
+  }
+
+  componentDidUpdate() {
+    this._renderCanvas();
+  }
```

Comment:
The canvas doesn't get rendered when the component updates (for instance, when a particular timeframe is zoomed on), that's why _renderCanvas had to be called in the componentDidUpdate() lifecycle method